### ISSUE TITLE
How to compute the J and JDotNu for contact vertices

### ISSUE THREAD

**Issue by VenusPasandi** [ami-iit/matlab-whole-body-simulator#67]
We need $J$ and $\dot{J} \nu$ of the contact vertices for computing the forward dynamics of the system. Actually, we need just the part of these terms that relates to the linear velocity and linear acceleration of the contact vertices.

Considering $f$ as the foot frame, we can write the linear velocity of a contact vertex of the frame $f$ like $i$ as:

\begin{equation}
v_i = v_f - S(r_{if}) \omega_f
\end{equation}

As $v = J_L \nu$ and $\omega_f = J_A \nu$ where $J_L$ and $J_A$ are the linear and angular parts of the Jacobian matrix of the frame $f$, we have

\begin{equation}
J_{L_i} = J_{L_f} - S(r_{if}) J_{A_f}
\end{equation}

Differentiating the above equation and multiplying it by $\nu$, we have

\begin{equation}
\dot{J}\_{L_i} \nu = \dot{J}\_{L_f} \nu - S(v_{if}) J_{A_f} \nu - S(r_{if}) \dot{J}\_{A_f} \nu
\end{equation}

However, in https://github.com/ami-iit/matlab-whole-body-simulator/blob/master/lib/%2Bmwbs/%40Contacts/compute_J_and_JDot_nu_in_contact_frames.m, we used the following relations:

\begin{equation}
J_{L_i} = J_{L_f} - S(r_{if}) J_{A_f}, \\
\dot{J}\_{L_i} \nu = \dot{J}\_{L_f} \nu - S(r_{if}) \dot{J}\_{A_f} \nu
\end{equation}

I can not understand why the term $- S(v_{if}) J_{A_f} \nu$ is ignored. This term is equal to $- (\omega_f \times r_{if}) \times \omega_f$. Does any body have any idea on it?

cc: @traversaro, @nunoguedelha, @GiulioRomualdi

**Comment by DanielePucci** [comment]
@VenusPasandi I believe that the math is not rendering properly 
![image](https://user-images.githubusercontent.com/5045846/160587402-1ecc5625-26bc-45e7-bf7c-1df7202a92a1.png)

**Comment by VenusPasandi** [comment]
> @VenusPasandi I believe that the math is not rendering properly ![image](https://user-images.githubusercontent.com/5045846/160587402-1ecc5625-26bc-45e7-bf7c-1df7202a92a1.png)

Sorry @DanielePucci. I tried to correct the rendering issue. It seems there is a rendering issue with writing $\dot{J}_i$!! I don't know if anyone has a solution for it?

cc: @ami-iit/artificial-mechanical-intelligence

**Comment by traversaro** [comment]
It is rendering fine for me. Anyhow, when you have a problem in some formula involving a `_`, you can tipically change it to `\_` to avoid the problems.

**Comment by VenusPasandi** [comment]
> It is rendering fine for me. Anyhow, when you have a problem in some formula involving a `_`, you can tipically change it to `\_` to avoid the problems.

Thank you @traversaro, your trick works nice :)

**Comment by traversaro** [comment]
> This term is equal to $- (\omega_f \times r_{if}) \times \omega_f$. Does any body have any idea on it?

Are you sure that that term does not evalute to 0 due to some formulas such as the one https://en.wikipedia.org/wiki/Triple_product ? Intuitively I would say now, but it may be worth to check ?

**Comment by S-Dafarra** [comment]
Is  $S(r_{if})$ changing? If $r_{if}$ is the relative position between the vertex and the center, then this would not change, right?

**Comment by VenusPasandi** [comment]
> Is S(rif) changing? If rif is the relative position between the vertex and the center, then this would not change, right?

Yes, $r_{if}$ is the relative position of the vertex in the foot frame. So, its magnitude is constant and does not change. But, its direction changes.

**Comment by S-Dafarra** [comment]
> > Is S(rif) changing? If rif is the relative position between the vertex and the center, then this would not change, right?
> 
> Yes, $r_{if}$ is the relative position of the vertex in the foot frame. So, its magnitude is constant and does not change. But, its direction changes.

It depends on the frame where it is measured. If it is measured in the frame attached to the foot, then also the direction is constant

**Comment by VenusPasandi** [comment]
> > > Is S(rif) changing? If rif is the relative position between the vertex and the center, then this would not change, right?
> > 
> > 
> > Yes, $r_{if}$ is the relative position of the vertex in the foot frame. So, its magnitude is constant and does not change. But, its direction changes.
> 
> It depends on the frame where it is measured. If it is measured in the frame attached to the foot, then also the direction is constant

Actually, the first time that I read the code, I thought the same as you.

But we are computing the time derivative with respect to the inertial/world frame. So, even if it is defined in the local frame, because of the rotation of the local frame, its time derivative is not  zero.

**Comment by Giulero** [comment]
Hi @VenusPasandi!

The math should be here https://github.com/ami-iit/matlab-whole-body-simulator/issues/1#issue-693202831

Maybe there is something wrong there. This is a good occasion to double check it! :)

**Comment by VenusPasandi** [comment]
> Hi @VenusPasandi!
> 
> The math should be here [#1 (comment)](https://github.com/ami-iit/matlab-whole-body-simulator/issues/1#issue-693202831)
> 
> Maybe there is something wrong there. This is a good occasion to double check it! :)

Hi @Giulero!
Thank you for your reference. Unfortunately, in https://github.com/ami-iit/matlab-whole-body-simulator/issues/1#issue-693202831 the part related to the JDotNu is ignored and only the final relation is written! But, there is a hint which says that we obtain the relation in the same way as the relation of J, so, not with the differentiation of the relation of $J$. So, let us try in this way: The acceleration of the point i is as

\begin{equation}
\begin{aligned}
a_i = a_f + \alpha_f \times r_{if} + \omega_f \times (\omega_f \times r_{if}) \\
= (\dot{J}\_{L_f} \nu + J_{L_f} \dot{\nu}) + (\dot{J}\_{A_f} \nu + J_{A_f} \dot{\nu}) \times r_{if}  -  (J_{A_f} \nu \times r_{if}) \times J_{A_f} \nu \\
= (\dot{J}\_{L_f} - S(r_{if}) \dot{J}\_{A_f}  - S(J_{A_f} \nu \times r_{if}) J_{A_f}) \nu + ( J_{L_f} - S(r_{if}) J_{A_f} ) \dot{\nu}
\end{aligned}
\end{equation}

So,

\begin{equation}
\begin{aligned}
J_i = J_{L_f} - S(r_{if}) J_{A_f} \\
\dot{J}\_i = \dot{J}\_{L_f} - S(r_{if}) \dot{J}\_{A_f}  - S(J_{A_f} \nu \times r_{if}) J_{A_f}
\end{aligned}
\end{equation}

So, again we have the term $S(J_{A_f} \nu \times r_{if}) J_{A_f}$ in the JDotNu expression. Note that, this term is the centrifugal acceleration of point i. So, if we don't have any assumption for the motion of the foot, in general, the centrifugal term is not zero.

**Comment by DanielePucci** [comment]
TLDR
>  This term is equal to $- (\omega_f \times r_{if}) \times \omega_f$. Does any body have any idea on it?

I do not know if it helps you @VenusPasandi, but notice that 

$- (\omega_f \times r_{if}) \times \omega_f = \omega_f \times (\omega_f \times r_{if}) = S(\omega_f )^2\times r_{if} = (\omega_f \omega^\top_f - |\omega_f|^2 I)r_{if}$

**Comment by VenusPasandi** [comment]
> TLDR
> 
> > This term is equal to $- (\omega_f \times r_{if}) \times \omega_f$. Does any body have any idea on it?
> 
> I do not know if it helps you @VenusPasandi, but notice that
> 
> $- (\omega_f \times r_{if}) \times \omega_f = \omega_f \times (\omega_f \times r_{if}) = S(\omega_f )^2r_{if} = (\omega_f \omega^\top_f - |\omega_f|^2 I)r_{if}$

Thank you @DanielePucci. This relation is really useful if we want to add this term to current implemented equation.

**Comment by VenusPasandi** [comment]
I couldn't find any reason/assumption explaining that the term $ \omega_f \times (\omega_f \times r_{if}) $ is equal to zero. So, to be sure, I tested this term in a simulation. I performed two simulations; first by ignoring this term and second by considering this term (where I implemented this term by using the formula that @DanielePucci mentioned in https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issuecomment-1082210907). The results are as the following where the red dotted curves denote the first case (ignoring this term) and the blue curves denote the second case (considering this term).

![image](https://user-images.githubusercontent.com/34647611/160806312-27b02acb-9a85-4262-9482-0eded39581aa.png)
![image](https://user-images.githubusercontent.com/34647611/160806352-3955172c-ea85-4b8d-bfd0-7d6f4d9cbd3d.png)
![image](https://user-images.githubusercontent.com/34647611/160806382-f94f4d49-0610-423d-9879-9381ff7bb8f3.png)
![image](https://user-images.githubusercontent.com/34647611/160806428-94620b7c-8e63-4135-a271-0b2ea46c8a44.png)
![image](https://user-images.githubusercontent.com/34647611/160806477-571d16cb-b2f9-48d0-9b46-1f7981fa5181.png)
![image](https://user-images.githubusercontent.com/34647611/160806528-8340900a-62f8-4b2b-8857-5b820223c2b3.png)
![image](https://user-images.githubusercontent.com/34647611/160806581-b563443b-02be-46f4-acfd-bfe450a71de8.png)
![image](https://user-images.githubusercontent.com/34647611/160806640-cbf18dd8-82aa-46f3-bf1b-0233dda0b0cd.png)
![image](https://user-images.githubusercontent.com/34647611/160806680-c8cb7608-3a14-43b0-8ac0-a82333db7071.png)

So, this term is not equal to zero and ignoring or considering it affects the robot velocity vector and contact wrenches. So, I think this term is ignored by a simple mistake like https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issuecomment-1081796461 if the reasoning explained in https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issue-1184596992 is correct.

So, I kindly ask @Giulero, @traversaro  and @DanielePucci  to check the math in https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issue-1184596992.

**Comment by Giulero** [comment]
@VenusPasandi! You're right!

I did the computations again. We know that:
$$
J_i \nu = \begin{bmatrix} I & -S({}^I R_\text{foot} {}^\text{foot} p_I) \end{bmatrix} J \nu
$$ 

so $J_i = \begin{bmatrix} I & -S({}^I R_\text{foot} {}^\text{foot} p_I) \end{bmatrix} J$.

Differentiating $J_i$ leads to:
$$
\dot{J_i} = \begin{bmatrix} I & -S({}^I R_\text{foot} {}^\text{foot} p_I) \end{bmatrix}  \dot{J}+ \begin{bmatrix} 0 & - S(S(J_A \nu) {}^I R_\text{foot} {}^\text{foot} p_I ) \end{bmatrix} J
$$

If we inject back $\nu$ and do some manipulations the following term appears:
$$
S(J_A \nu) S(J_A \nu) {}^I R_\text{foot} {}^\text{foot} p_I ,
$$
that's what you call $\omega \times \omega \times r$.

Now my main concern is that $\nu$ is an optimization variable, and with this new term we lose the linear relationship with $\nu$. It seems that using https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issuecomment-1082210907 we can extract a quadratic relationship, but how do we compute the 2-norm?

**Comment by Giulero** [comment]
Ah, I missed your comment! :D 
How do you write the optimization problem?

**Comment by VenusPasandi** [comment]
Thank you @Giulero 

> Now my main concern is that $\nu$ is an optimization variable, and with this new term we lose the linear relationship with $\nu$. It seems that using https://github.com/ami-iit/matlab-whole-body-simulator/issues/67#issuecomment-1082210907 we can extract a quadratic relationship,

Actually, the time derivative of $\nu$ i.e $\dot{\nu}$ is the optimisation variable. If you remember in https://github.com/ami-iit/matlab-whole-body-simulator/issues/1#issue-693202831, we have $\dot{J} \nu + J \dot{\nu}$ where $\dot{\nu}$ is an optimisation variable. So, we don't need the linear relationship with respect to $\nu$.

**Comment by Giulero** [comment]
Yes, I went through the math and the code and my memory went back! :) 
Thank you!

**Comment by VenusPasandi** [comment]
Thank you all for the nice discussion. I added the missing term into the dynamics block via #71

**Comment by DanielePucci** [comment]
Very well done @VenusPasandi and @Giulero !

**Comment by diegoferigo** [comment]
Ignore this comment, necessary to render math with [![green-pi](https://img.shields.io/badge/Rendered%20with-Green%20Pi-00d5b1?style=flat-square)](https://github.com/nschloe/green-pi?activate&inlineMath=$)
